bits/BIT-0008-EVM-Weight-Verification.md

```diff
@@ -0,0 +1,417 @@
+        // --- End Verification Logic ---
+
+        // Calculate the hash of the weights using abi.encode for unambiguous serialization.
+        // The Subtensor (Rust) side must replicate this exact encoding.
+        bytes32 weightsHash = keccak256(abi.encode(weights));
+
+        // Call the precompile to store the hash.
+        // The precompile validates the caller (`msg.sender` must be this contract) and
+        // the transaction origin (`tx.origin` must be the selected validator).
+        bool success = i_subtensor.storeVerifiedWeightsHash(i_netuid, weightsHash);
+        require(success, "Precompile call failed");
+
+        emit WeightsVerified(tx.origin, i_netuid, weightsHash);
+    }
+}
+```
+
+> [!IMPORTANT]
+> The hashing mechanism uses `abi.encode(weights)` for unambiguous serialization. The Subtensor runtime must replicate this exact ABI encoding format. This includes the 32-byte offset header and 32-byte length prefix that `abi.encode` adds for dynamic arrays.
+
+## Security Considerations
+
+### Transfer of Power
+
+Enforcement of the verification contract transfers power from the validator to the subnet owner. This is a significant change from the current system, where the validator is not held responsible for the validity of their weights by subnet owners directly.
+
+To outline some extreme cases where the subnet owner could abuse the verification contract:
+
+1. The subnet owner can craft a malicious verification contract that is designed to fail for a specific validator.
+2. The subnet owner can create a flawed contract that results in unexpected rejections for valid submissions.
+3. The subnet owner could add a constraint to the contract that forces validators to pay a fee or perform an additional task to verify their weights.
+
+To mitigate these risks, a separate governance proposal is required to manage the lifecycle and approval of verification contracts. This is outside the scope of this BIT.
+
+### Smart Contract Vulnerabilities
+
+Subnet owners are responsible for defining their own verification contracts, and as such, they are responsible for the security of their contracts. If incorrectly implemented, these contracts could result in vulnerabilities such as:
+
+1. Flawed RBAC (Role-Based Access Control) allowing for unauthorized access to the verification contract. While the [#precompile] prevents non-validators from submitting weight verifications, it should not be possible for non-validators to
+2. Incorrect handling of the `data` parameter, which could result in arbitrary code execution.
+
+### `drand` Dependency Risk
+
+The optimistic validity verification mechanism relies on the `drand` randomness beacon to select the validator to submit a weight verification. If the `drand` beacon is compromised, the validator selection process could be manipulated to favor certain validators. Note that this risk is shared with the current system, where the [commit-reveal] process is also dependent on `drand`.
```

Comment:
if no drand beacons for a full tempo, lets fall back to block hash (in case c-r period is 2 tempo and we lost drand a tempo ago, we can still operate, or when set_weights is in use instead of drand)